pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -165,9 +164,11 @@ public long getLatestIngestionTimestamp() {
           dictionaryColumnSize = dataType.size();
         }
         String allocationContext = buildAllocationContext(_segmentName, column, V1Constants.Dict.FILE_EXTENSION);
+        // NOTE: preserve 20% buffer for cardinality to reduce the chance of re-sizing the dictionary
+        int estimatedCardinality = (int) (_statsHistory.getEstimatedCardinality(column) * 1.2);
```

Comment:
I think 10% should be sufficient, but then I agree that both 10 and 20 percent are arbitrary. Why not make this a per-machine config? Default can be 10, and we can play with it as needed.

Reply:
@Jackie-Jiang I think we should merge this change soon. Do you plan to add a per0-machine config in this PR or later? Or change to 10%?

Reply:
I can change it to 10%. The long term solution should be expanding the buffers as more values added. Allocating huge buffer at once is a quite slow operation (hundreds milliseconds), which should be avoided.